src/samplers/cpu/cpi/mod.rs

```diff
@@ -91,17 +179,86 @@ impl Sampler for Cpi {
+        let mut cur_group_data = vec![];
+        for group in &mut self.groups {
+            match group.read_group() {
+                Ok(counts) => {
+                    let running_us;
+                    match (counts.time_enabled(), counts.time_running()) {
+                        (Some(enable_time), Some(running_time)) => {
+                            if enable_time.as_nanos() != running_time.as_nanos() {
+                                error!("The perf group {} is not always running", group.cpu);
+                                continue;
+                            }
+                            running_us = running_time.as_micros()
+                        }
+                        (_, _) => {
+                            error!(
+                                "The perf group {} has no enabled time and running time",
+                                group.cpu
+                            );
+                            continue;
+                        }
+                    }
+                    if counts.time_enabled() != counts.time_running() {}
+                    let cycles = counts[&(group.cycles)];
+                    let instructions = counts[&(group.instructions)];
+
+                    total_cycles += cycles;
+                    total_instructions += instructions;
+                    let tsc = counts[&(group.tsc)];
+                    base_frequency = tsc / running_us as u64;
+                    BASEFREQUENCY.set(base_frequency as i64);
+                    cur_group_data.push(counts);
+                }
+                Err(e) => {
+                    error!("error in sampling the perf group: {e}");
+                }
+            };
+        }
+
+        if cur_group_data.len() == self.groups.len() {
+            // update total cycles and instructions
+            self.total_cycles.set(now, elapsed, total_cycles);
+            self.total_instructions
+                .set(now, elapsed, total_instructions);
+            // update the IPC and IPNS
+            if let Some(prev_group_data) = &self.prev_group_data {
+                let nr_cpu = cur_group_data.len() as i64;
+                if cur_group_data.len() == prev_group_data.len() {
+                    let mut ipc: i64 = 0;
+                    let mut avg_frequency = 0;
+                    let mut ipns: i64 = 0;
+                    for cpu in 0..cur_group_data.len() {
+                        let cur_cycles = cur_group_data[cpu][&(self.groups[cpu].cycles)];
+                        let prev_cycles = prev_group_data[cpu][&(self.groups[cpu].cycles)];
+                        let cur_instructions =
+                            cur_group_data[cpu][&(self.groups[cpu].instructions)];
+                        let prev_instructions =
+                            prev_group_data[cpu][&(self.groups[cpu].instructions)];
+                        let cpu_ipc = (cur_instructions - prev_instructions) * 1000 / (cur_cycles - prev_cycles);
```

Comment:
@brayniac These are all computing per-CPU metrics, so once Rezolus supports per-CPU metrics, we can expose them easily.